db/dcrpg/queries.go

```diff
@@ -3619,6 +3620,116 @@ func appendBlockFees(charts *cache.ChartData, rows *sql.Rows) error {
+// cache.ChartUpdater.
+func retrievePrivacyParticipation(ctx context.Context, db *sql.DB, charts *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectMixedTotalPerBlock, charts.TotalMixedTip())
+	if err != nil {
+		return nil, err
+	}
+	return rows, nil
+}
+
+// Append the result from retrievePrivacyParticipation to the provided
+// ChartData. This is the Appender half of a pair that make up a
+// cache.ChartUpdater.
+func appendPrivacyParticipation(charts *cache.ChartData, rows *sql.Rows) error {
+	defer rows.Close()
+	blocks := charts.Blocks
+	for rows.Next() {
+		var blockHeight uint64
+		var totalMixed int64
+		if err := rows.Scan(&blockHeight, &totalMixed); err != nil {
+			log.Errorf("Unable to scan for MixedVoutsPerBlock fields: %v", err)
+			return err
+		}
+		if totalMixed < 0 {
+			totalMixed *= -1
+		}
+
+		// Converting to atoms.
+		blocks.TotalMixed = append(blocks.TotalMixed, uint64(totalMixed))
+	}
+	return rows.Err()
+}
+
+// retrieveAnonymitySet retrieves any block total mixed data that is newer than
+// the data in the provided ChartData. This data is used to plot anonymity-set
+// on the /charts page. This is the Fetcher half of a pair that make up a
+// cache.ChartUpdater.
+func retrieveAnonymitySet(ctx context.Context, db *sql.DB, _ *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectMixedVouts)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+// Append the result from retrieveAnonymitySet to the provided ChartData. This
+// is the Appender half of a pair that make up a cache.ChartUpdater.
+func appendAnonymitySet(charts *cache.ChartData, rows *sql.Rows) (err error) {
+	var vals, fundHeights, spendHeights []int64
+
+	var maxHeight int64
+	minHeight := int64(math.MaxInt64)
+
+	for rows.Next() {
+		var value, fundHeight, spendHeight int64
+		var spendHeightNull sql.NullInt64
+		var tree uint8
+		err = rows.Scan(&value, &fundHeight, &spendHeightNull, &tree)
+		if err != nil {
+			return err
+		}
+
+		if spendHeightNull.Valid {
+			spendHeight = spendHeightNull.Int64
+		} else {
+			spendHeight = -1
+		}
+
+		if fundHeight < minHeight {
+			minHeight = fundHeight
+		}
+
+		if spendHeight > maxHeight {
+			maxHeight = spendHeight
+		}
+
+		vals = append(vals, value)
+		fundHeights = append(fundHeights, fundHeight)
+		spendHeights = append(spendHeights, spendHeight)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Errorf("Unable to scan for MixedVouts fields: %v", err)
+		return err
+	}
+
+	blocks := charts.Blocks
+
+	// block anonymity set
+	for h := len(blocks.AnonymitySet) + 1; h <= len(blocks.Height); h++ {
+		var anonymitySet int64
+		for iu := range vals {
+			if int64(h) >= fundHeights[iu] && (int64(h) < spendHeights[iu] || spendHeights[iu] == -1) {
+				anonymitySet += vals[iu]
+			}
+		}
+		blocks.AnonymitySet = append(blocks.AnonymitySet, uint64(anonymitySet))
+	}
```

Comment:
Unfortunately, I don't think this approach to computing the anonymity set *at each height* is scalable.  Even at the current height and mixing status, it takes far too long.  Rather than compute anon set value fresh at each height, please try to do this by adding and removing value from a running anonymity set value depending on what heights are new for the current height.  Feel free to use different data structures as you Scan data from the db (e.g. map by fund height and map by spend height?  but please give it additional though before trying that).

Reply:
What I have done here is to use a running total and maintain a map of spending heights for all the values. At every new height, deduct all spent vouts from the running total and append the total as the anonymity set for that height

Reply:
I just saw https://github.com/decred/dcrdata/pull/1640/commits/e3e2d6e3a72e2ec3eca0462409da97ca9e2e9526 addressing this.  Looks to be what I had in mind.  Thanks.